input/common/linux_common.c

```diff
@@ -121,3 +156,222 @@ bool linux_terminal_disable_input(void)
 
    return true;
 }
+
+static void linux_poll_illuminance_sensor(void *data)
+{
+   linux_illuminance_sensor_t *sensor = data;
+
+   if (!data)
+      return;
+
+   while (!sensor->done)
+   { /* Aligned int reads are atomic on most CPUs */
+        double lux;
+        int millilux;
+        unsigned poll_rate = sensor->poll_rate;
+
+        /* Don't allow cancellation inside the critical section,
+         * as it opens up a file; we don't want to leak it! */
+        pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, NULL);
+        lux = linux_read_illuminance_sensor(sensor);
+        millilux = (int)(lux * 1000.0);
+        retro_assert(poll_rate != 0);
+
+        sensor->millilux = millilux;
+        pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, NULL);
+
+        /* Allow cancellation here so that the main thread doesn't block
+         * while waiting for this thread to wake up and exit. */
+        retro_sleep(1000 / poll_rate);
+        if (errno == EINTR)
+        {
+           RARCH_ERR("Illuminance sensor thread interrupted\n");
+           break;
+        }
+   }
+
+   RARCH_DBG("Illuminance sensor thread for %s exiting\n", sensor->path);
+}
+
+linux_illuminance_sensor_t *linux_open_illuminance_sensor(unsigned rate)
+{
+   DIR *devices = NULL;
+   struct dirent *d = NULL;
+   linux_illuminance_sensor_t *sensor = malloc(sizeof(*sensor));
+
+   if (!sensor)
+      goto error;
+
+   sensor->millilux  = 0;
+   sensor->poll_rate = rate ? rate : DEFAULT_POLL_RATE;
+   sensor->thread    = NULL; /* We'll spawn a thread later, once we find a sensor */
+   sensor->done      = false;
+
+   devices = opendir(IIO_DEVICES_DIR);
+   if (!devices)
+   { /* If we couldn't find the IIO device directory... */
+      char errmesg[NAME_MAX_LENGTH];
+      strerror_r(errno, errmesg, sizeof(errmesg));
```

Comment:
I don't see us ever using this before in the entire codebase. Why is this done and what is its purpose? Can you not just use strlcpy instead?

Reply:
I'm not sure I understand, "this" being what exactly? My use of `strerror_r`?

Reply:
Yes, why is this necessary?

Reply:
[`opendir(3)`](https://man7.org/linux/man-pages/man3/opendir.3.html) sets `errno` upon failure, and [`strerror_r(3)`](https://man7.org/linux/man-pages/man3/strerror.3.html) is thread-safe.

Reply:
Don't we have wrapper functions for the likes of opendir in libretro-common? I'd prefer we just go through that instead. retro_dirent IIRC.

Reply:
Ah, yes, I'd forgotten about those. I'll switch to them, not a problem.